### PR TITLE
Reader monad

### PR DESCRIPTION
**Pull request by mpurland**
A reader monad with some simple test cases

### CODE REVIEW

**Comment by CodaFi** [comment]
I believe the test cases in the [Writer Spec](https://github.com/typelift/Swiftz/blob/master/SwiftzTests/WriterSpec.swift) should port over with very little resistance.

File: Swiftz/Reader.swift [hunk 1]
```diff
@@ -0,0 +1,149 @@
+//
+//  Reader.swift
+//  Swiftz
+//
+//  Created by Matthew Purland on 11/25/15.
+//  Copyright © 2015 TypeLift. All rights reserved.
+//
+
+import Foundation
```
Comment (CodaFi) [inline Swiftz/Reader.swift]: Remove this

File: Swiftz/Reader.swift [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+/// A type for a `Reader` monad
+public protocol ReaderType {
```
Comment (CodaFi) [inline Swiftz/Reader.swift]: This is an interesting approach to MonadReader, but I don't think it gets us much.  Still, I think there's something here that can be applied to the codebase as a whole in the future.  Can you draw up an issue discussing this style?

File: Swiftz/Reader.swift [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    init(_ reader : R -> A) {
+        f = reader
+    }
+    
+    public func mapReader<B>(f : A -> B) -> Reader<R, B> {
+        return Reader<R, B>(f • runReader)
+    }
+}
+
+extension Reader: ReaderType {
```
Comment (CodaFi) [inline Swiftz/Reader.swift]: TypeLift colons, pls

File: Swiftz/Reader.swift [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+extension Reader : Monad {
+    public func bind(f : A -> Reader<R, B>) -> Reader<R, B> {
+        return self >>- f
+    }
+}
+
+func >>- <R, A, B>(r : Reader<R, A>, f : A -> Reader<R, B>) -> Reader<R, B> {
+    return Reader<R, B>.init({ (environment: R) -> B in
```
Comment (CodaFi) [inline Swiftz/Reader.swift]: Should be

```
return Reader<R, B> { (environment : R) -> B in
        let a = r.runReader(environment)
        let readerB = f(a)
        let b = readerB.runReader(environment)
        return b
}

**Comment by CodaFi** [comment]
It's missing the global functions [ask](http://hackage.haskell.org/package/mtl-2.2.1/docs/src/Control-Monad-Reader-Class.html#ask) and [asks](http://hackage.haskell.org/package/mtl-2.2.1/docs/src/Control-Monad-Reader-Class.html#asks).

File: Swiftz/Reader.swift (hunk 2 above)
Comment (mpurland) [inline Swiftz/Reader.swift]: I think in many cases protocols are the way to go, but may not be needed in some places.

Comment (CodaFi) [inline Swiftz/Reader.swift]: They definitely are.  But what I want is to introduce them all at once into a future major version of the framework.  For now, I think we should hold off on introducing them into just one place and plan for bigger things.

Comment (mpurland) [inline Swiftz/Reader.swift]: Sounds like a plan. :+1:

**Comment by mpurland** [comment]
@CodaFi Any thoughts about the ask/asks api? I tried to match as closely as I could to Haskell mtl in Swift.

File: Swiftz/Reader.swift [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+public func runReader<R, A>(reader : () -> R -> A) -> R -> A {
+    return runReader(reader())
+}
+
+public func runReader<R, A>(reader : Reader<R, A>) -> R -> A {
+    return reader.runReader
+}
+
+/// Retrieves the monad environment
+public func ask<A>() -> A -> A {
```
Comment (CodaFi) [inline Swiftz/Reader.swift]: This should just be a monadic action in the reader monad.

```swift
public func ask<A>() -> Reader<A> { //... }
```

Comment (mpurland) [inline Swiftz/Reader.swift]: This was supposed to be just a helper, but can be removed.

Comment (CodaFi) [inline Swiftz/Reader.swift]: I've never seen this particular overloading before.  What's the expected use case?

Comment (mpurland) [inline Swiftz/Reader.swift]: It was intended to give ask functionality of the form A -> A to support runReader. 

```swift
let lengthResult3 = runReader(ask())(1234)
XCTAssert(lengthResult3 == 1234)
        
let lengthResult4 = runReader(ask)(1234)
XCTAssert(lengthResult4 == 1234)
```

I'm open to suggestions. How would you update the ```ask``` function to be a monadic action in the reader? So far I've run through a few different options, but each isn't working as I saw it should with runReader.

I've been working through some options, but none of these are right.

```swift
public static func ask<A>() -> (R -> A) -> Reader<R, A> {
    return { (f : R -> A) in
        return Reader<R, A>(f)
    }
}

public static func ask<A>() -> A -> Reader<R, A> {
    return { a in
        return Reader<R, A> { _ in a }
    }
}

public static func ask<A>(f : () -> A) -> Reader<A, A> {
    return Reader<A, A> { _ in f() }
}
```

Comment (mpurland) [inline Swiftz/Reader.swift]: @CodaFi Should I remove the global ask function then? What do you think?

Comment (CodaFi) [inline Swiftz/Reader.swift]: I think it's subsumed by just passing. identity

**Comment by mpurland** [comment]
@CodaFi Latest commit with update based on your comments.

**Comment by CodaFi** [comment]
:+1: Give it the old rebase and we're good to go

**Comment by mpurland** [comment]
Rebased from master now. After tests pass should be all ready to go. :)

**Comment by CodaFi** [comment]
:fire: 

:boat:

**Pull request merged**